### ISSUE TITLE
How to debug gimic

### ISSUE THREAD

**Issue by mariavd** [qmcurrents/gimic#101]
There was an unusual issue with reading a very large (4.5 GB) XDENS file on our local cluster, which does not occur on the other cluster and on my computer, so I would like to investigate what is going on. I was thinking of using gdb but there is some misunderstanding between the pythons. I realise that this is not on your priority list right now but do you know how I can get gdb to process gimic? Or is there another way to debug the code? I compiled it with the -g flag from inside the virtual environment and, of course, the first line in the python script of the `gimic` script is 

```
#!/home/mariavd/venv/bin/python
```

But then: 

```
(venv)[mariavd@gold integration]$ gdb /home/mariavd/venv/bin/python /home/mariavd/kevin-test/gimic/build/bin/gimic gimic.inp
'import site' failed; use -v for traceback
Traceback (most recent call last):
  File "<string>", line 32, in <module>
  File "/opt/rh/python27/root/usr/lib64/python2.7/os.py", line 400, in <module>
    import UserDict
  File "/opt/rh/python27/root/usr/lib64/python2.7/UserDict.py", line 116, in <module>
    import _abcoll
  File "/opt/rh/python27/root/usr/lib64/python2.7/_abcoll.py", line 456, in <module>
    KeysView.register(type({}.viewkeys()))
AttributeError: 'dict' object has no attribute 'viewkeys'
Excess command line arguments ignored. (gimic.inp)
GNU gdb (GDB) Red Hat Enterprise Linux (7.2-92.el6)
Copyright (C) 2010 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
```

**Comment by bast** [comment]
Without the gdb, does it crash immediately?

**Comment by mariavd** [comment]
No, this temporary Gimic works fine, and also the tests completed successfully.

**Comment by mariavd** [comment]
Hmm, it is actually a problem with gdb. When I don't give any program as a parameter, it still lists those python errors.

**Comment by bast** [comment]
I mean reading the large file which crashes on your cluster, does it crash immediately?

**Comment by mariavd** [comment]
No, it starts and some 5-10 seconds later it crashes:

```
 Thu Dec  7 16:30:29 2017
  TITLE:

 INFO: Detected TURBOMOLE input

 Number of atoms = 500

 Normalizing basis

   Total number of primitive  GTO's ****
   Total number of contracted GTO's 7500

 *** Calculating screening coefficients
 INFO: Screening threshold:   0.1000E-07

At line 83 of file /home/mariavd/gimic/src/libgimic/dens.f90 (unit = 43, file = '../XDENS')
Fortran runtime error: End of file
```

The only difference between the crashing and working Gimic is the compiler (ifort for the working and gfortran for the crashing one) but I guess this is not related to the problem.

**Comment by bast** [comment]
https://github.com/qmcurrents/gimic/blob/master/src/libgimic/dens.f90#L83

**Comment by bast** [comment]
I would in this case not go for a debugger but I would write a little 20 line Fortran code that does exactly what the above line does, reading exactly the same file and check whether that crashes too.

**Comment by mariavd** [comment]
After discussing with Radovan, I am posting several details of what I tried regarding reading the XDENS file. It might be useful for future reference. 
```
$ wc -l XDENS
225000003 XDENS
```

When reading the XDENS file with a simple Fortran loop:

```
do i=1,(readLines)
     read(XDFD,*) L
end do
```

For `readLines = 225000000` the file is read properly. Asking for one more line causes the program to crash:

```
$ gfortran readXDENS.f90 && ./a.out 
   225000001
At line 22 of file readXDENS.f90 (unit = 30, file = 'XDENS')
Fortran runtime error: End of file

Backtrace for this error:
#0  0x0000003360cac794 in __libc_wait (stat_loc=0x0) at ../sysdeps/unix/sysv/linux/wait.c:32
#1  0x000000336341400d in ?? () from /usr/lib64/libgfortran.so.3
#2  0x000000336341582e in ?? () from /usr/lib64/libgfortran.so.3
#3  0x0000003363415a81 in _gfortran_generate_error () from /usr/lib64/libgfortran.so.3
#4  0x00000033634b8689 in ?? () from /usr/lib64/libgfortran.so.3
#5  0x00000033634b8b3c in ?? () from /usr/lib64/libgfortran.so.3
#6  0x0000000000400c63 in MAIN__ ()
#7  0x0000000000400d71 in main ()
```

Since there is nothing unusual about the file and that the calculation was running successfully on two more computers, it is very unlikely to be a problem with Gimic. The XDENS looks fine. 

```
$ tail XDENS
0.56754001114947E-02
0.79971626666072E-02
-.22868921150965E-01
0.57862972318246E-01
-.85234769463374E-03
0.10739442019430E-02
-.22159650730928E-03
0.37165698984348E-05
-.29552227433960E-02
0.00000000000000E+00
```

Radovan suggested that this could be caused by a limit in the number of lines that can be read. I tried the INQUIRE command to check the record length but the output was not useful. It returns -1 unless the record length is specified at the OPEN command. This will require further investigation.

**Comment by bast** [comment]
I think this has been solved in #162.
